cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/DefaultApplicationSpecification.java

```diff
@@ -185,4 +187,14 @@ public ArtifactId getArtifactId() {
   public Map<String, Plugin> getPlugins() {
     return plugins;
   }
+
+  @Override
+  public Set<String> getProgramsByType(ProgramType programType) {
+    return programSpecificationsByProgramType.get(programType).keySet();
+  }
+
+  private static <T extends ProgramSpecification> Map<String, T> convertValues(
+    Map<String, ProgramSpecification> programSpecificationMap, Class<T> specificationClass) {
+    return programSpecificationMap.entrySet().stream().collect(Collectors.toMap(e -> e.getKey(), e-> (T)e.getValue()));
```

Comment:
You should cast it with the provided class:

```
Collectors.toMap(Map.Entry::getKey, specClass::cast)
```

Reply:
Thanks!

Reply:
Actually I don't think that works since we need to cast the value, not the entire entryset